clang/lib/AST/ItaniumMangle.cpp

```diff
@@ -2285,7 +2285,7 @@ void CXXNameMangler::mangleQualifiers(Qualifiers Quals, const DependentAddressSp
     if (Context.getASTContext().addressSpaceMapManglingFor(AS)) {
       //  <target-addrspace> ::= "AS" <address-space-number>
       unsigned TargetAS = Context.getASTContext().getTargetAddressSpace(AS);
-      if (TargetAS != 0)
+      if (TargetAS != 0 || (Context.getASTContext().getLangOpts().SYCLIsDevice))
```

Comment:
`sycl_private` address is mapped to `0` for SPIR as well as default address space, so in order to "specialize" templates with the `sycl_private` qualified types, we have to mangle them differently from the unqualified types.
Test case is here: https://github.com/intel/llvm/pull/968/files#diff-4e74dfb66665848ae647af85422c78d8R41

Reply:
> sycl_private address is mapped to 0 for SPIR as well as default address space

Default address space is supposed to be akin to generic, i.e. it should be mapped to 4.

Reply:
Sorry, my previous reply is not quite accurate.
This change is needed because pointers qualified with address space mapped to `0` LLVM AS are mangled the same way as unquliafied pointers.

Reply:
Ok, that makes sense.
However, I'm surprised that we don't have a similar code for OpenCL here. `opencl_private` is also mapped to 0, but it has a distinct mangling:
```
__attribute__((overloadable))
void foo(__private int *ip) {
  *ip = 0;
}
// magled as: _Z3fooPU9CLprivatei
```

Reply:
We have similar problem for OpenCL targeting **SPIR**. You seemed to use some other target your example. X86?

Reply:
Sorry, the problem I'm talking about is mangling, not the mangled names conflict.
Mangled names conflict is not a problem in OpenCL as OpenCL compiler qualifies all the pointers.

Reply:
> We have similar problem for OpenCL targeting **SPIR**. You seemed to use some other target your example. X86?

Right, sorry for the confusion. With `-triple spir` the function is mangled as `_Z3fooPi`.

> Sorry, the problem I'm talking about is mangling, not the mangled names conflict.
> Mangled names conflict is not a problem in OpenCL as OpenCL compiler qualifies all the pointers.

I don't quite understand what the problem is. Can provide a test case?

Reply:
The test case is added to this PR already: https://github.com/intel/llvm/pull/968/files#diff-9cbf615457fd73dff7e9840ed855fa9a and problem is covered in this thread. I just clarified why it affects only SYCL compiler, but not OpenCL.
Smaller reproducer:
```c++
template<typename T>
void tmpl(T t){}

__attribute__((opencl_private)) int *PRIV;
tmpl(PRIV);

int *NoAS;
tmpl(NoAS);
```
SYCL compiler produces the same mangled name for two `tmpl` instances, OpenCL compiler adds `generic` qualifier for `NoAS` declaration and produces different names.
Let me know if it's still unclear.

I think original patch from @erichkeane mapped `sycl_private` address space to AS `5`. This might be an alternative option, but we need LLVM-SPIRV translator to support alternative mappings as well.